src/ucp/core/ucp_worker.c

```diff
@@ -2350,3 +2350,109 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
 
     UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(worker);
 }
+
+static unsigned ucp_worker_discard_uct_ep_destroy_progress(void *arg)
+{
+    ucp_request_t *req  = (ucp_request_t*)arg;
+    uct_ep_h uct_ep     = req->send.discard_uct_ep.uct_ep;
+    ucp_worker_h worker = req->send.discard_uct_ep.ucp_worker;
+
+    ucs_debug("destroy uct_ep=%p", uct_ep);
+    uct_ep_destroy(uct_ep);
+    ucp_request_put(req);
+    --worker->flush_ops_count;
+
+    return 1;
+}
+
+static void
+ucp_worker_discard_uct_ep_flush_comp(uct_completion_t *self,
+                                     ucs_status_t status)
+{
+    uct_worker_cb_id_t cb_id    = UCS_CALLBACKQ_ID_NULL;
+    ucp_request_t *req          = ucs_container_of(self, ucp_request_t,
+                                                   send.state.uct_comp);
+
+    ucs_trace_req("flush completion req=%p status=%d", req, status);
+
+    /* don't destroy UCT EP from the flush completion callback,
+     * schedule a progress callback to destroy UCT EP */
+    uct_worker_progress_register_safe(req->send.discard_uct_ep.ucp_worker->uct,
+                                      ucp_worker_discard_uct_ep_destroy_progress,
+                                      req, UCS_CALLBACKQ_FLAG_ONESHOT, &cb_id);
+}
+
+static unsigned ucp_worker_discard_uct_ep_progress(void *arg)
+{
+    ucp_request_t *req = (ucp_request_t*)arg;
+    uct_ep_h uct_ep    = req->send.discard_uct_ep.uct_ep;
+    ucs_status_t status;
+
+    req->send.state.uct_comp.func  = ucp_worker_discard_uct_ep_flush_comp;
+    req->send.state.uct_comp.count = 1;
+
+    for (;;) {
```

Comment:
how does it make sense without calling progress? I think need to reschedule `ucp_worker_discard_uct_ep_progress` for next try.
also pls make code more explicit, e.g.:
```
if (status == UCS_UCS_ERR_BUSY) {
    /* try again code */
}
```

Reply:
if `uct_ep_pending_add()` completest with `UCS_UCS_ERR_BUSY`, then `uct_ep_flush()` will return either `UCS_INPROGRESS` or `UCS_OK`, that;s why we do `for(;;)` to make `uct_ep_flush()` again

Reply:
you mean UCS_NO_RESOURCES from flush and ERR_BUSY from pending_add are mutually exclusive cases?

Reply:
> you mean UCS_NO_RESOURCES from flush and ERR_BUSY from pending_add are mutually exclusive cases?

yes, if `uct_ep_flush()` returns `UCS_ERR_NO_RESOURCES`, and then `uct_ep_pending_add()` returns `UCS_UCS_ERR_BUSY`, we could expect then a subsequent `uct_ep_flush()` call should be completed with `UCS_OK`/`UCS_INPROGRESS`

Reply:
how is it possible without [async]progress?

Reply:
> how is it possible without [async]progress?

yes, this is possible only in case of async-progress.
in other cases, adding to pending queue should return success/in-progress after NO_RESOURCES from `uct_ep_flush()`

Reply:
pending add can't return IN_PROGRESS :)
IMHO better to keep code simpler for non-fast-path, also this case might happen very rare...

Reply:
> pending add can't return IN_PROGRESS :)

yes, I meant `UCS_OK`
@evgeny-leksikov do you mean to re-register progress cb on UCT worker's callback queue and then retry?
@brminich @yosefe wdyt?

Reply:
BTW, rescheduling one-shot-GC-progress in resource critical state we can keep resources free for something eager

Reply:
> @evgeny-leksikov do you mean to re-register progress cb on UCT worker's callback queue and then retry?

yes, try flush/pending_add one time per CB invoke

Reply:
> BTW, rescheduling one-shot-GC-progress in resource critical state we can keep resources free for something eager

yes, make sense to me